File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.java
// 최단 거리 문제 - 특정 시작점에서 여러 위치까지의 최단 거리 계산
// 격자나 그래프에서 BFS를 활용하여 최단 거리 구하기
// BFS를 사용하면 가중치가 동일한 경우, 최단 거리를 쉽게 찾음

import java.util.*;

class Solution {
    
    int[] dr = {-1,1,0,0};
    int[] dc = {0,0,-1,1};
    int n,m;
    boolean[][] visited;
    
    public int solution(int[][] maps) {
        
        n = maps.length;
        m = maps[0].length;
        
        visited = new boolean[n][m];
        Queue<int[]> q = new ArrayDeque<>();
        
        q.offer(new int[]{0,0,1}); // r 좌표, c 좌표, 이동한 거리 세팅
        visited[0][0] = true;
        
        while(!q.isEmpty()) {
            int[] cur = q.poll();
            int r = cur[0];
            int c = cur[1];
            int dist = cur[2];
            
            if(r == n - 1 && c == m - 1) { // 목적지인지 먼저 확인
                return dist;
            }
            
            for(int i = 0; i < 4; i++) {
                int nr = r + dr[i];
                int nc = c + dc[i];
                
                if(nr < 0 || nr >= n || nc < 0 || nc >= m) continue;
                if(visited[nr][nc] || maps[nr][nc] == 0) continue;
                
                visited[nr][nc] = true;
                q.offer(new int[]{nr, nc, dist + 1});
            
            }
        }
        return -1;
    }
}

File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.py
from collections import deque

def solution(maps):
    n = len(maps)
    m = len(maps[0])
    visited = [[False] * m for _ in range(n)]
    queue = deque()
    
    # ✅ BFS 탐색을 수행한다.
    queue.append((0, 0, 1))
    visited[0][0] = True
    
    dr = [0, 1, 0, -1]
    dc = [-1, 0, 1, 0]
    
    while queue:
        r, c, dist = queue.popleft()
        
        # ✅ 탐색 과정에서 목적지에 도달하면 거리를 반환한다.
        if r == n - 1 and c == m - 1:
            return dist
        
        for d in range(4):
            nr = r + dr[d]
            nc = c + dc[d]
            
            if 0 <= nr < n and 0 <= nc < m and maps[nr][nc] == 1:
                if not visited[nr][nc]:
                    visited[nr][nc] = True
                    queue.append((nr, nc, dist + 1))
    
    # ✅ 탐색 과정에서 목적지에 도달하지 못했다면 -1을 반환한다.
    return -1